File: cortical_create_normative_data_from_tsf.m
function [lh_M, rh_M] = cortical_create_normative_data_from_tsf()

SUBJECTS_DIR = getenv('SUBJECTS_DIR');


f_subjects = [SUBJECTS_DIR '/templates/subjects_to_average.txt'];

metrics = {'fa','md','ad','rd',...
           'afd-par','afd-perp'};

subjects = readlines(f_subjects);
subjects = subjects(strlength(strtrim(subjects)) > 0);

lh_tsfs = containers.Map();
rh_tsfs = containers.Map();

for m = 1:numel(metrics)
    metric = metrics{m};
    lh_tsfs(metric) = cortical_find_subject_files(SUBJECTS_DIR, subjects, ['lh_ico6_sym_' metric '.tsf']);
    rh_tsfs(metric) = cortical_find_subject_files(SUBJECTS_DIR, subjects, ['rh_ico6_sym_' metric '.tsf']);
end

% Figure out the number of vertices so and build a large
% multidimensional matrix
dummyvar = lh_tsfs(metrics{1});
onetsf   = dummyvar{1};
dummytsf = read_mrtrix_tsf(onetsf);
nVerts   = size(dummytsf.data,2);
nMetrics = length(metrics);
dummyMat = cortical_cell2mat(dummytsf.data);
nDepths  = size(dummyMat,2) .* 2; % pad it twice as large, as other subjects may have deeper streamlines.
nSubjects= length(lh_tsfs(metrics{1}));
lh_M        = nan(nVerts,nDepths,nSubjects,nMetrics);
rh_M        = nan(nVerts,nDepths,nSubjects,nMetrics);

% Average across subjects, per vertex/streamline and per depth point.
% Streamline lengths vary within and across subjects (cortical thickness),
% so each subject's data is NaN-padded to a common width before averaging
% with 'omitnan', which naturally accounts for the varying number of
% subjects contributing at each depth point.
lh_avg = containers.Map();
rh_avg = containers.Map();
lh_std = containers.Map();
rh_std = containers.Map();
lh_n   = containers.Map();
rh_n   = containers.Map();

% Also save the full stack for each metric so we can build a
% multidimensional array
lh_stack = containers.Map();
rh_stack = containers.Map();

out_dir = fullfile(SUBJECTS_DIR, 'templates', 'normative');
if ~exist(out_dir, 'dir')
    mkdir(out_dir);
end

for m = 1:numel(metrics)
    metric = metrics{m};
    fprintf(1,'  Averaging %s (LH n=%d, RH n=%d)\n', metric, ...
        numel(lh_tsfs(metric)), numel(rh_tsfs(metric)));

    [lh_avg(metric), lh_std(metric), lh_n(metric), lh_stack(metric)] = cortical_average_tsf_files(lh_tsfs(metric));
    [rh_avg(metric), rh_std(metric), rh_n(metric), rh_stack(metric)] = cortical_average_tsf_files(rh_tsfs(metric));

    l_nDepths     = size(lh_stack(metric),2);
    r_nDepths     = size(rh_stack(metric),2);
    lh_M(:,1:l_nDepths,:,m) = lh_stack(metric);
    rh_M(:,1:r_nDepths ,:,m) = rh_stack(metric);

    save_tsf_matrix(lh_avg(metric), fullfile(out_dir, ['lh_ico6_sym_' metric '_mean.tsf']));
    save_tsf_matrix(lh_std(metric), fullfile(out_dir, ['lh_ico6_sym_' metric '_std.tsf']));
    save_tsf_matrix(lh_n(metric),   fullfile(out_dir, ['lh_ico6_sym_' metric '_n.tsf']));

    save_tsf_matrix(rh_avg(metric), fullfile(out_dir, ['rh_ico6_sym_' metric '_mean.tsf']));
    save_tsf_matrix(rh_std(metric), fullfile(out_dir, ['rh_ico6_sym_' metric '_std.tsf']));
    save_tsf_matrix(rh_n(metric),   fullfile(out_dir, ['rh_ico6_sym_' metric '_n.tsf']));

    cortical_save_gifti_matrix(lh_avg(metric), fullfile(out_dir, ['lh_ico6_sym_' metric '_mean.func.gii']));
    cortical_save_gifti_matrix(lh_std(metric), fullfile(out_dir, ['lh_ico6_sym_' metric '_std.func.gii']));
    cortical_save_gifti_matrix(lh_n(metric),   fullfile(out_dir, ['lh_ico6_sym_' metric '_n.func.gii']));

    cortical_save_gifti_matrix(rh_avg(metric), fullfile(out_dir, ['rh_ico6_sym_' metric '_mean.func.gii']));
    cortical_save_gifti_matrix(rh_std(metric), fullfile(out_dir, ['rh_ico6_sym_' metric '_std.func.gii']));
    cortical_save_gifti_matrix(rh_n(metric),   fullfile(out_dir, ['rh_ico6_sym_' metric '_n.func.gii']));
end


% remove the extra Nans at the end of second dimension in the large
% multidimensional matrices
lh_toRemove = sum(isnan(lh_M(:,:,1,1))) == nVerts;
lh_M(:,lh_toRemove,:,:) = [];
rh_toRemove = sum(isnan(rh_M(:,:,1,1))) == nVerts;
rh_M(:,rh_toRemove,:,:) = [];

fprintf(1,'Sizes of lh_M and rh_M:\n')
fprintf(1,'lh_M: %s\n', mat2str(size(lh_M)));
fprintf(1,'rh_M: %s\n', mat2str(size(rh_M)));


% save a .mat file for easier data handling of multivariate data
mat_fname = fullfile(out_dir, ['multivariate.mat']);
fprintf(1,'Saving multivariate mat file: %s\n',mat_fname);
save(mat_fname,'lh_M','rh_M','metrics','subjects');


function save_tsf_matrix(M, filename)
% Write a NaN-padded (per-streamline, per-depth-point) matrix to a
% MRtrix .tsf file, trimming each row's trailing NaN padding back to
% that streamline's real length before writing.

tsf = struct();
tsf.data = trim_nan_padding(M);
write_mrtrix_tsf(tsf, filename);
fprintf(1,'    Wrote %s\n', filename);

end


function data = trim_nan_padding(M)
% Convert an N x maxLen NaN-padded matrix (NaN padding only ever at the
% end of each row, per cortical_cell2mat's convention) back into a
% 1 x N cell array of column vectors with the padding removed.

nRows = size(M, 1);
data = cell(1, nRows);
for i = 1:nRows
    valid = find(~isnan(M(i, :)), 1, 'last');
    if isempty(valid)
        valid = 0;
    end
    data{i} = M(i, 1:valid).';
end

end

end% function

File: cortical_create_normative_data_from_tsf.py
#!/usr/bin/env python3
"""
cortical_create_normative_data_from_tsf.py — Python port of
cortical_create_normative_data_from_tsf.m

Builds a normative (control-cohort) dataset from per-subject MRtrix .tsf
files and saves it as a single HDF5 file for use by the cortical DWI viewer's
future normative-comparison / multivariate (Mahalanobis) explorer features.

Unlike the MATLAB version, this does NOT also precompute/save per-metric
mean/std/n — those are cheaply derivable from the raw per-subject stack via
np.nanmean/np.nanstd/count-of-non-NaN, so storing them separately would just
be redundant derived data that could drift out of sync. Consumers should
compute them on demand from lh_M/rh_M.

Also unlike the MATLAB version, a subject's per-metric contribution is placed
at a FIXED index (its position in subjects_to_average.txt) rather than being
appended positionally as files are found. This keeps the subject axis
consistently aligned across metrics even if some subjects are missing a
particular metric's file — the MATLAB script implicitly assumes every
subject has every metric, which would silently misalign the subject axis
across the metric dimension if that assumption doesn't hold.

Usage:
    python cortical_create_normative_data_from_tsf.py [subjects_dir]

If subjects_dir is omitted, the SUBJECTS_DIR environment variable is used
(matching the MATLAB script's getenv('SUBJECTS_DIR')).
"""
import os, sys, glob, argparse
import numpy as np
import h5py

sys.path.insert(0, os.path.dirname(__file__))
from cortical_io import read_mrtrix_tsf, pad_to_matrix
from cortical_browser_config import TEMPLATE, METRICS   # shared with cortical_browser.py


def find_one_subject_file(subjects_dir, subject, filename):
    """Recursively search one subject's directory for a file with the given
    exact basename (mirrors cortical_find_subject_files.m, but per-subject
    so a missing file can be reported/handled individually)."""
    matches = glob.glob(os.path.join(subjects_dir, subject, '**', filename), recursive=True)
    return matches[0] if matches else None


def load_subject_matrix(tsf_path):
    """One subject's (nVerts, ownMaxLen) matrix, -1 sentinel masked to NaN
    (mirrors cortical_cell2mat(..., 'MaskInvalid', true))."""
    _, tracks = read_mrtrix_tsf(tsf_path)
    M = pad_to_matrix(tracks)
    M[M == -1] = np.nan
    return M


def stack_subjects(subjects_dir, subjects, filename):
    """(nVerts, maxDepth, nSubjects) NaN-padded stack for one metric/hemi,
    aligned to `subjects` by index. A subject missing this file contributes
    an all-NaN slice, keeping the subject axis consistent across metrics."""
    per_subj = [None] * len(subjects)
    n_found = 0
    for i, subj in enumerate(subjects):
        path = find_one_subject_file(subjects_dir, subj, filename)
        if path is not None:
            per_subj[i] = load_subject_matrix(path)
            n_found += 1
    print(f'    {filename}: {n_found}/{len(subjects)} subjects found')

    n_verts = next((m.shape[0] for m in per_subj if m is not None), 0)
    max_len = max((m.shape[1] for m in per_subj if m is not None), default=0)
    stack = np.full((n_verts, max_len, len(subjects)), np.nan, dtype=np.float32)
    for s, M in enumerate(per_subj):
        if M is not None:
            stack[:, :M.shape[1], s] = M
    return stack


def build_normative_stack(subjects_dir, subjects, metrics, hemi, template=TEMPLATE):
    """(nVerts, nDepths, nSubjects, nMetrics) stack for one hemisphere,
    across all metrics, NaN-padded to the deepest metric's depth and
    trimmed of any trailing depth columns that are all-NaN everywhere."""
    per_metric_stacks = []
    for metric in metrics:
        filename = f'{hemi}_{template}_{metric}.tsf'
        per_metric_stacks.append(stack_subjects(subjects_dir, subjects, filename))

    n_verts   = per_metric_stacks[0].shape[0]
    n_subj    = len(subjects)
    n_depths  = max(s.shape[1] for s in per_metric_stacks)
    n_metrics = len(metrics)
    M = np.full((n_verts, n_depths, n_subj, n_metrics), np.nan, dtype=np.float32)
    for m, stack in enumerate(per_metric_stacks):
        M[:, :stack.shape[1], :, m] = stack

    keep = ~np.all(np.isnan(M), axis=(0, 2, 3))
    return M[:, keep, :, :]


def main():
    ap = argparse.ArgumentParser(description='Precompute an HDF5 normative dataset from a subject cohort')
    ap.add_argument('subjects_dir', nargs='?', default=os.environ.get('SUBJECTS_DIR'))
    args = ap.parse_args()
    if not args.subjects_dir:
        sys.exit('subjects_dir not given and SUBJECTS_DIR is not set')

    subjects_file = os.path.join(args.subjects_dir, 'templates', 'subjects_to_average.txt')
    if not os.path.isfile(subjects_file):
        sys.exit(f'Subject list not found: {subjects_file}')
    with open(subjects_file) as f:
        subjects = [line.strip() for line in f if line.strip()]
    print(f'Cohort  : {len(subjects)} subjects from {subjects_file}')
    print(f'Metrics : {METRICS}')

    out_dir = os.path.join(args.subjects_dir, 'templates', 'normative')
    os.makedirs(out_dir, exist_ok=True)
    out_path = os.path.join(out_dir, f'{TEMPLATE}_multivariate.h5')

    str_dtype = h5py.string_dtype(encoding='utf-8')
    with h5py.File(out_path, 'w') as h5f:
        h5f.attrs['template'] = TEMPLATE
        h5f.create_dataset('metrics',  data=np.array(METRICS,  dtype=object), dtype=str_dtype)
        h5f.create_dataset('subjects', data=np.array(subjects, dtype=object), dtype=str_dtype)
        for hemi in ('lh', 'rh'):
            print(f'Building {hemi}_M …')
            M = build_normative_stack(args.subjects_dir, subjects, METRICS, hemi)
            print(f'  {hemi}_M shape: {M.shape}  (nVerts, nDepths, nSubjects, nMetrics)')
            h5f.create_dataset(f'{hemi}_M', data=M, compression='gzip', compression_opts=4)

    size_mb = os.path.getsize(out_path) / 1e6
    print(f'\nSaved {out_path} ({size_mb:.1f} MB)')


if __name__ == '__main__':
    main()
